octobot_commons/databases/bases/base_database.py

```diff
@@ -0,0 +1,94 @@
+#  Lesser General Public License for more details.
+#
+#  You should have received a copy of the GNU Lesser General Public
+#  License along with this library.
+import contextlib
+import numpy
+
+import octobot_commons.databases.document_database_adaptors as adaptors
+import octobot_commons.databases.bases.document_database as document_database
+import octobot_commons.databases.database_caches as database_cache
+
+
+class BaseDatabase:
+    def __init__(self, file_path: str, database_adaptor=adaptors.TinyDBAdaptor, cache_size=None, **kwargs):
+        if database_adaptor is not None:
+            self._database = document_database.DocumentDatabase(database_adaptor(file_path, cache_size=cache_size,
+                                                                                 **kwargs))
+            self._database.initialize()
+        self.are_data_initialized = False
+        self.are_data_initialized_by_key = {}
+        self.cache = database_cache.GenericDatabaseCache()
+
+    def set_initialized_flags(self, value, keys=None):
+        self.are_data_initialized = value
+        for key in keys or self.are_data_initialized_by_key.keys():
+            self.are_data_initialized_by_key[key] = value
+
+    def get_db_path(self):
+        return self._database.get_db_path()
+
+    async def search(self, dict_query: dict = None):
+        if dict_query is None:
+            return await self._database.query_factory()
+        return (await self._database.query_factory()).fragment(dict_query)
+
+    async def count(self, table_name: str, query) -> int:
+        return await self._database.count(table_name, query)
+
+    async def flush(self):
+        await self._database.flush()
+
+    async def hard_reset(self):
+        return await self._database.hard_reset()
+
+    async def close(self):
+        await self.flush()
+        await self._database.close()
+
+    async def clear(self):
+        self.cache.clear()
+
+    async def contains_row(self, table: str, row: dict):
+        if self.cache.contains_row(table, row):
+            return True
+        return await self.count(table, await self.search(row)) > 0
+
+    def __str__(self):
+        return f"{self.__class__.__name__}, database: {self._database}"
+
+    @classmethod
+    @contextlib.asynccontextmanager
+    async def database(cls, *args, with_lock=False, cache_size=None, database_adaptor=None, **kwargs):
+        if with_lock:
+            adaptor = kwargs.pop("database_adaptor", adaptors.TinyDBAdaptor)
+            if adaptor is None:
+                raise RuntimeError("database_adaptor parameter required")
+            adaptor_instance = adaptor(*args, cache_size=cache_size, **kwargs)
+            database = cls(*args, database_adaptor=database_adaptor, cache_size=cache_size, **kwargs)
+            async with document_database.DocumentDatabase.locked_database(adaptor_instance) as db:
+                database._database = db
+                yield database
```

Comment:
I think we should put this into a `_create_database_instance` method or something like this to be clearer